packages/@uppy/utils/src/fetcher.js

```diff
@@ -0,0 +1,156 @@
+/* eslint-disable no-param-reassign */
+/* eslint-disable max-len */
+
+import NetworkError from './NetworkError.js'
+import ProgressTimeout from './ProgressTimeout.js'
+
+const noop = () => {}
+
+/**
+ * Get an abort signal that will be aborted when the specified Uppy file is removed or all uploads are cancelled.
+ *
+ * @param {Uppy} uppy
+ *  The Uppy instance.
+ * @param {string} [id]
+ *  The ID of the file to watch for removal.
+ * @param {AbortSignal} [additionalSignal]
+ *  An optional additional abort signal.
+ * @returns {AbortController}
+ *  The abort signal.
+ */
+export function getUppyAbortController(uppy, id, additionalSignal) {
+  const controller = new AbortController()
+
+  uppy.once('cancel-all', () => {
+    controller.abort()
+  })
+
+  if (id) {
+    uppy.on('file-removed', (file) => {
+      if (id === file.id) {
+        controller.abort()
+      }
+    })
+  }
+
+  if (additionalSignal) {
+    additionalSignal.addEventListener('abort', () => {
+      controller.abort()
+    })
+  }
+
+  return controller
+}
+
+/**
+ * Fetches data from a specified URL using XMLHttpRequest, with optional retry functionality and progress tracking.
+ *
+ * @param {string} url
+ *  The URL to send the request to.
+ * @param {object} [options]
+ *  Optional settings for the fetch operation.
+ * @param {string} [options.method='GET']
+ *  The HTTP method to use for the request.
+ * @param {string} [options.body=null]
+ *  The request payload, if any.
+ * @param {string} [options.timeout=30000]
+ *  Miliseconds between XMLHttpRequest upload progress events before the request is aborted.
+ * @param {string} [options.withCredentials=false]
+ *  Sets the withCredentials property of the XMLHttpRequest object.
+ * @param {string} [options.responseType='']
+ *  Sets the responseType property of the XMLHttpRequest object.
+ * @param {Record<string, string>} [options.headers]
+ *  An object representing any headers to send with the request.
+ * @param {number} [options.retries=3]
+ *  The number of retry attempts to make if the request fails.
+ * @param {number} [options.delay=1000]
+ *  The initial delay between retry attempts, in milliseconds. The delay doubles with each retry.
+ * @param {function(Event): void} [options.onUploadProgress]
+ *  A callback function for tracking upload progress.
+ * @param {function(XMLHttpRequest): boolean} [options.shouldRetry]
+ *  A function to determine whether to retry the request.
+ * @param {function(): void} [options.onTimeout]
+ *  Called when when no XMLHttpRequest upload progress events have been received for `timeout` ms.
+ * @param {AbortSignal} [options.signal]
+ *  signal to abort the upload.
+ * @returns {Promise<XMLHttpRequest>}
+ *  A Promise that resolves to the response text if the request succeeds, and rejects with an error if it fails.
+ */
+export function fetcher(url, options = {}) {
```

Comment:
I see that you tried to use `ky` first but we need upload progress which is not supported. I think if we are implementing our own request abstraction, ideally it should have unit tests because it becomes such an important core component of Uppy. Could we alternatively look into using [`axios`](https://github.com/axios/axios) instead of implementing our own wrapper? axios is extremely popular and seems to support upload events too.

Reply:
I prefer our own abstraction over a library, especially since it's mostly just a promise wrapper. But your right about tests!